consensus/types/src/beacon_state.rs

```diff
@@ -250,7 +250,7 @@ where
 
     // Light-client sync committees
     #[superstruct(only(Altair))]
-    pub current_sync_committee: SyncCommittee<T>,
+    pub current_sync_committee: Arc<SyncCommittee<T>>,
```

Comment:
I don't think we want to do an `Arc` here, it means that whenever we clone `BeaconState` both the original and the clone will point to the same `SyncCommittee`. Whilst it *seems* nice to reduce copies of `SyncCommittee`, the reference-counting code is quite complex and adds significant overhead (it involves atomics).

I'd be interested to hear the issue you are solving here.

Reply:
The goal was to reduce `SyncCommittee` clones in gossip verification specifically, because we'll need it for every non-duplicate sync committee message we receive every slot.  I think the napkin math is: 48 bytes per pubkey * 1024 pubkeys per committee * 1024 messages per slot = 48MB.

Initially I was looking into creating a cache in the beacon chain, similar to the `ShufflingCache`. But that became tricky because the shuffling decision root for sync committees can end up being further than `SLOTS_PER_HISTORICAL_ROOTS` in the past.  

I spoke to @michaelsproul about it, and we were thinking adding an `Arc` could be an alternative. But maybe the simplest solution is best here and we should just clone.

Reply:
Forgot that `SYNC_COMMITTEE_SIZE` has been reduced from 1024 to 512 in the latest spec, so that's 4x more of a reason to just use clones

Reply:
I'm a big fan of this `Arc` because of the cloning optimisations it enables. The sync committees are 100% immutable after they're computed and they're shared between multiple `BeaconState`s (as `current_` and `next_`) for a really long time. That's exactly the use-case for an `Arc`. The slowness of `Arc` is also only relative to regular memory dereferencing: I think it would be faster (perhaps insignificantly) in this context.

In favour of `Arc`:

* Faster cloning: cloning an `Arc` involves a handful of cycles to increment the atomic refcount (see [clone](https://doc.rust-lang.org/src/alloc/sync.rs.html#1267)), which I think _has to be_* faster than allocating and then copying (even for the ~24KB sync committees).
* No overhead on dereferencing (see [deref](https://doc.rust-lang.org/src/alloc/sync.rs.html#1299-1306)).
* Less cloning of the real data: less total memory usage, less fragmentation, no temporary allocations.
* Even with the slightly strange `ssz::Decode` impl that ignores sharing, it degrades gracefully to have similar memory properties to if we'd cloned the underlying data all along.

Against:

* This is a bit of an unknown, as we don't have any other `Arc`s in consensus code, and there might be unexpected performance gotchas that we're not aware of?
* As asterisked above*, I haven't actually benchmarked to see if it's faster.

My other motivation for wanting to get an `Arc` into the `BeaconState` is as a trial for putting _more_ `Arc`s in there :sweat_smile:. As we move to higher validator counts I think minimising the amount of cloning we do of full `BeaconState` objects with `Vec`s of 200K-1M validators will be important. Being able to share data between states using `Arc`s could greatly reduce the total amount we store in memory, and the total time it takes to clone a state. This is the approach taken by Proto's [remerkleable](https://github.com/protolambda/remerkleable) library (thanks to Python's GC), and the approach I experimented with at EthDenver 2020 with [`ValidatorTree`](https://github.com/michaelsproul/lighthouse/blob/d501b7e3796206611593111123941f6cc45e1571/eth2/types/src/beacon_state/validator_tree.rs#L32-L41).

If you poke around that `ValidatorTree` you'll see I did the completely insane thing of having an `Arc` and an `RwLock` for every node in the tree. Even with this naive approach, the state transition was not that much slower (~30-40%), and cloning was like 10x faster. The main pain point in the scheme was iterating the `validators` registry, which went from being an `O(n)` vec iteration to a `O(n log n)` tree traversal (ouch). One day I want to experiment with chunking the tree in different ways to make iteration faster, or using a structure more like the tree hash cache with the nodes in layers (plus `Arc`s!).

Reply:
> I'm a big fan of this Arc because of the cloning optimisations it enables

This is a good point, if it's intentional that we want to go down this path them I'm also OK with it.

I mistakenly assumed this was a once-off decision rather than a strategic move. My apologies for that.

Reply:
All good!